File: core-ops/go/harness.go
// Command amber-core-ops-go measures the exported operations of the Go
// Amber-Store core in process, through its library APIs.
//
// The driver is deliberately small: it owns fixture construction, a case
// registry and a timing loop, and writes one JSON document with the shared
// raw-sample schema (see ../SCHEMA.md). Every counterpart case in the Rust
// driver (../rust) carries the same op and workload names, so the report
// pairs them without any per-language table.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"sort"
	"time"
)

// ---------------------------------------------------------------------------
// Case registry
// ---------------------------------------------------------------------------

// Case is one measured operation at one workload. Setup builds whatever the
// operation reads; only Run is timed. Ops is the number of core operations
// performed by one Run call — short operations are batched so the measured
// interval stays far above the clock's resolution — and Bytes is the payload
// the operation moved, where a throughput figure is meaningful (0 otherwise).
type Case struct {
	Group    string
	Op       string
	Workload string
	// Threads records the concurrency the operation was asked for: 1 for a
	// strictly single-threaded call, n for an explicit worker count, and 0
	// for an operation that picks its own parallelism (GOMAXPROCS here,
	// available_parallelism in Rust — both bounded by the CPU set the
	// driver runs under).
	Threads int
	Ops     int
	Bytes   int64
	// PerRep re-runs Setup before every repetition. Destructive operations
	// (compaction, wipe, repair, removal) need a fresh copy each time; the
	// copy is made outside the measured interval.
	PerRep bool
	Setup  func(*Env) any
	Run    func(*Env, any) uint64
	Free   func(*Env, any)
}

// Check is one correctness assertion. Digest, when set, is a canonical
// fingerprint of the operation's output; the report compares digests of the
// same check id across the two cores, which is how format agreement is
// established rather than assumed.
type Check struct {
	ID     string `json:"id"`
	Group  string `json:"group"`
	Op     string `json:"op"`
	Passed bool   `json:"passed"`
	Detail string `json:"detail,omitempty"`
	Digest string `json:"digest,omitempty"`
	// Comparable says whether the digest is supposed to be equal in the
	// other core. It is false where the two implementations are documented
	// to be interoperable without being byte-identical — zstd-compressed
	// record payloads are produced by klauspost/compress in Go and libzstd
	// in Rust, so records, segment bodies and wire packs containing them
	// differ. Such a digest is still recorded, as a within-core regression
	// anchor; the report never fails on a mismatch it was told not to
	// compare.
	Comparable bool `json:"comparable"`
}

// Unsupported records an exported operation that exists in one core only, or
// a workload the core cannot express. It never produces a sample: an absent
// operation must not be readable as an infinitely fast one.
type Unsupported struct {
	Op       string `json:"op"`
	Workload string `json:"workload,omitempty"`
	Reason   string `json:"reason"`
}

// Sample is one repetition of one case.
type Sample struct {
	Group     string `json:"group"`
	Op        string `json:"op"`
	Workload  string `json:"workload"`
	Threads   int    `json:"threads"`
	Rep       int    `json:"rep"`
	Ops       int    `json:"ops"`
	Bytes     int64  `json:"bytes"`
	WallNs    int64  `json:"wall_ns"`
	CPUUserNs int64  `json:"cpu_user_ns"`
	CPUSysNs  int64  `json:"cpu_sys_ns"`
	MaxRSSKiB int64  `json:"max_rss_kib"`
	Status    string `json:"status"`
}

// Counter is a runtime-specific number that has no comparable counterpart in
// the other core. The report prints these in their own table and never uses
// them in a paired comparison.
type Counter struct {
	Op       string `json:"op"`
	Workload string `json:"workload"`
	Rep      int    `json:"rep"`
	Name     string `json:"name"`
	Value    int64  `json:"value"`
}

// Env is everything the cases share: the profile, the scratch directory and
// the fixtures built once at startup.
type Env struct {
	Profile Profile
	Scratch string
	Xattrs  bool
	Sink    uint64

	fx *Fixtures

	checks      []Check
	samples     []Sample
	counters    []Counter
	unsupported []Unsupported
}

func (e *Env) check(c Check) { e.checks = append(e.checks, c) }

// pass records a passing check whose digest must equal the other core's.
func (e *Env) pass(group, op, id, digest string) {
	e.check(Check{ID: id, Group: group, Op: op, Passed: true, Digest: digest, Comparable: true})
}

// passLocal records a passing check whose digest is implementation-specific
// and must not be compared across cores.
func (e *Env) passLocal(group, op, id, digest string) {
	e.check(Check{ID: id, Group: group, Op: op, Passed: true, Digest: digest, Comparable: false})
}

// fail records a failing check. A failing check makes the whole run invalid;
// the report refuses to compare anything from an invalid run.
func (e *Env) fail(group, op, id, detail string) {
	e.check(Check{ID: id, Group: group, Op: op, Passed: false, Detail: detail})
}

// want records a check whose outcome is a boolean the caller computed.
func (e *Env) want(group, op, id string, ok bool, detail, digest string) {
	if ok {
		e.pass(group, op, id, digest)
		return
	}
	e.fail(group, op, id, detail)
}

// wantLocal is want for an implementation-specific digest.
func (e *Env) wantLocal(group, op, id string, ok bool, detail, digest string) {
	if ok {
		e.passLocal(group, op, id, digest)
		return
	}
	e.fail(group, op, id, detail)
}

func (e *Env) skip(op, workload, reason string) {
	e.unsupported = append(e.unsupported, Unsupported{Op: op, Workload: workload, Reason: reason})
}

// ---------------------------------------------------------------------------
// The timing loop
// ---------------------------------------------------------------------------

// run executes one case: Warmup unmeasured repetitions followed by Reps
// measured ones. Setup (and, for destructive cases, the per-repetition
// rebuild) stays outside every measured interval.
func (e *Env) run(c Case) {
	if c.Ops <= 0 {
		c.Ops = 1
	}
	var state any
	if !c.PerRep {
		state = c.Setup(e)
	}
	reset := func() any {
		if c.PerRep {
			return c.Setup(e)
		}
		return state
	}
	release := func(s any) {
		if c.PerRep && c.Free != nil {
			c.Free(e, s)
		}
	}

	for i := 0; i < e.Profile.Warmup; i++ {
		s := reset()
		e.Sink += c.Run(e, s)
		release(s)
	}
	for rep := 0; rep < e.Profile.Reps; rep++ {
		s := reset()
		runtime.GC()
		var m0, m1 runtime.MemStats
		runtime.ReadMemStats(&m0)
		u0, s0 := cpuTime()
		t0 := time.Now()
		e.Sink += c.Run(e, s)
		wall := time.Since(t0)
		u1, s1 := cpuTime()
		runtime.ReadMemStats(&m1)
		e.samples = append(e.samples, Sample{
			Group: c.Group, Op: c.Op, Workload: c.Workload, Threads: c.Threads,
			Rep: rep, Ops: c.Ops, Bytes: c.Bytes,
			WallNs:    wall.Nanoseconds(),
			CPUUserNs: u1 - u0, CPUSysNs: s1 - s0,
			MaxRSSKiB: maxRSS(),
			Status:    "ok",
		})
		e.counters = append(e.counters,
			Counter{c.Op, c.Workload, rep, "go.heap_alloc_bytes", int64(m1.TotalAlloc - m0.TotalAlloc)},
			Counter{c.Op, c.Workload, rep, "go.mallocs", int64(m1.Mallocs - m0.Mallocs)},
		)
		release(s)
	}
	if !c.PerRep && c.Free != nil {
		c.Free(e, state)
	}
}

// runAll executes the registry in a stable order so two runs of the same
// profile touch the machine in the same sequence.
func (e *Env) runAll(cases []Case) {
	sort.SliceStable(cases, func(i, j int) bool {
		if cases[i].Group != cases[j].Group {
			return cases[i].Group < cases[j].Group
		}
		if cases[i].Op != cases[j].Op {
			return cases[i].Op < cases[j].Op
		}
		return cases[i].Workload < cases[j].Workload
	})
	for _, c := range cases {
		started := time.Now()
		e.run(c)
		if os.Getenv("AMBER_CORE_OPS_TRACE") != "" {
			fmt.Fprintf(os.Stderr, "  %-34s %-28s %8.1f ms\n", c.Op, c.Workload,
				float64(time.Since(started).Nanoseconds())/1e6)
		}
	}
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

type report struct {
	Schema      string        `json:"schema"`
	Core        string        `json:"core"`
	Profile     string        `json:"profile"`
	Config      Profile       `json:"config"`
	Identity    Identity      `json:"identity"`
	Environment Environment   `json:"environment"`
	StartedAt   string        `json:"started_at"`
	FinishedAt  string        `json:"finished_at"`
	Checks      []Check       `json:"checks"`
	Samples     []Sample      `json:"samples"`
	Counters    []Counter     `json:"counters"`
	Unsupported []Unsupported `json:"unsupported"`
	Blackhole   uint64        `json:"blackhole"`
}

func (e *Env) write(path string, id Identity, env Environment, started, finished time.Time) error {
	r := report{
		Schema: sampleSchema, Core: "go", Profile: e.Profile.Name,
		Config: e.Profile, Identity: id, Environment: env,
		StartedAt:  started.UTC().Format(time.RFC3339Nano),
		FinishedAt: finished.UTC().Format(time.RFC3339Nano),
		Checks:     e.checks, Samples: e.samples, Counters: e.counters,
		Unsupported: e.unsupported, Blackhole: e.Sink,
	}
	if r.Checks == nil {
		r.Checks = []Check{}
	}
	if r.Samples == nil {
		r.Samples = []Sample{}
	}
	if r.Counters == nil {
		r.Counters = []Counter{}
	}
	if r.Unsupported == nil {
		r.Unsupported = []Unsupported{}
	}
	b, err := json.MarshalIndent(r, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, append(b, '\n'), 0o644)
}

const sampleSchema = "amber-core-ops/samples/1"

File: core-ops/rust/src/harness.rs
//! The case registry, the timing loop and the sample schema.
//!
//! This mirrors `../go/harness.go` field for field: both drivers write the
//! same JSON document, so the report pairs Go and Rust results by `(op,
//! workload)` without knowing anything about either language.

use std::any::Any;
use std::time::Instant;

use serde::Serialize;

use crate::env::{Env, Profile};

/// State a case's setup hands to its run function. Setup is never timed.
pub type State = Box<dyn Any>;

/// Builds the state a case reads. Never inside a measured interval.
pub type SetupFn = Box<dyn Fn(&Env) -> State>;

/// The measured call. Returns a value the driver accumulates, so the work
/// cannot be optimised away.
pub type RunFn = Box<dyn Fn(&Env, &mut State) -> u64>;

/// Releases per-repetition state. Never inside a measured interval.
pub type FreeFn = Box<dyn Fn(&Env, State)>;

/// One measured operation at one workload.
pub struct Case {
    pub group: &'static str,
    pub op: String,
    pub workload: String,
    /// The concurrency the operation was asked for: 1 for a strictly
    /// single-threaded call, n for an explicit worker count, and 0 for an
    /// operation that picks its own parallelism.
    pub threads: usize,
    /// Core operations performed by one `run` call. Short operations are
    /// batched so the measured interval stays far above the clock.
    pub ops: usize,
    /// Payload bytes moved per `run` call; 0 where no throughput figure is
    /// meaningful.
    pub bytes: i64,
    /// Rebuild the state before every repetition. Destructive operations
    /// need a fresh copy each time; making it stays outside the interval.
    pub per_rep: bool,
    pub setup: SetupFn,
    pub run: RunFn,
    pub free: Option<FreeFn>,
}

impl Case {
    pub fn new(
        group: &'static str,
        op: &str,
        workload: &str,
        threads: usize,
        ops: usize,
        setup: SetupFn,
        run: RunFn,
    ) -> Case {
        Case {
            group,
            op: op.to_string(),
            workload: workload.to_string(),
            threads,
            ops: ops.max(1),
            bytes: 0,
            per_rep: false,
            setup,
            run,
            free: None,
        }
    }

    pub fn bytes(mut self, n: i64) -> Case {
        self.bytes = n;
        self
    }

    pub fn per_rep(mut self, free: FreeFn) -> Case {
        self.per_rep = true;
        self.free = Some(free);
        self
    }
}

/// A correctness assertion. `digest`, when set, is a canonical fingerprint of
/// the operation's output; the report compares digests of the same check id
/// across the two cores, which is how format agreement is established rather
/// than assumed.
#[derive(Debug, Clone, Serialize)]
pub struct Check {
    pub id: String,
    pub group: String,
    pub op: String,
    pub passed: bool,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub detail: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub digest: String,
    /// Whether the digest is supposed to equal the other core's. It is false
    /// where the two implementations are documented to be interoperable
    /// without being byte-identical — zstd-compressed record payloads come
    /// from libzstd here and klauspost/compress in Go, so records, segment
    /// bodies and wire packs containing them differ.
    pub comparable: bool,
}

/// An exported operation that exists in one core only, or a workload the core
/// cannot express. It never produces a sample: an absent operation must not
/// be readable as an infinitely fast one.
#[derive(Debug, Clone, Serialize)]
pub struct Unsupported {
    pub op: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub workload: String,
    pub reason: String,
}

/// One repetition of one case.
#[derive(Debug, Clone, Serialize)]
pub struct Sample {
    pub group: String,
    pub op: String,
    pub workload: String,
    pub threads: usize,
    pub rep: usize,
    pub ops: usize,
    pub bytes: i64,
    pub wall_ns: i64,
    pub cpu_user_ns: i64,
    pub cpu_sys_ns: i64,
    pub max_rss_kib: i64,
    pub status: String,
}

/// A runtime-specific number with no comparable counterpart in the other
/// core. The report prints these in their own table.
#[derive(Debug, Clone, Serialize)]
pub struct Counter {
    pub op: String,
    pub workload: String,
    pub rep: usize,
    pub name: String,
    pub value: i64,
}

/// Everything a run accumulates.
#[derive(Default)]
pub struct Recorder {
    pub checks: Vec<Check>,
    pub samples: Vec<Sample>,
    pub counters: Vec<Counter>,
    pub unsupported: Vec<Unsupported>,
    pub sink: u64,
}

impl Recorder {
    /// Records a passing check whose digest must equal the other core's.
    pub fn pass(&mut self, group: &str, op: &str, id: &str, digest: impl Into<String>) {
        self.checks.push(Check {
            id: id.to_string(),
            group: group.to_string(),
            op: op.to_string(),
            passed: true,
            detail: String::new(),
            digest: digest.into(),
            comparable: true,
        });
    }

    /// Records a passing check whose digest is implementation-specific and
    /// must not be compared across cores.
    pub fn pass_local(&mut self, group: &str, op: &str, id: &str, digest: impl Into<String>) {
        self.checks.push(Check {
            id: id.to_string(),
            group: group.to_string(),
            op: op.to_string(),
            passed: true,
            detail: String::new(),
            digest: digest.into(),
            comparable: false,
        });
    }

    pub fn fail(&mut self, group: &str, op: &str, id: &str, detail: impl Into<String>) {
        self.checks.push(Check {
            id: id.to_string(),
            group: group.to_string(),
            op: op.to_string(),
            passed: false,
            detail: detail.into(),
            digest: String::new(),
            comparable: true,
        });
    }

    pub fn want(
        &mut self,
        group: &str,
        op: &str,
        id: &str,
        ok: bool,
        detail: impl Into<String>,
        digest: impl Into<String>,
    ) {
        if ok {
            self.pass(group, op, id, digest);
        } else {
            self.fail(group, op, id, detail);
        }
    }

    pub fn want_local(
        &mut self,
        group: &str,
        op: &str,
        id: &str,
        ok: bool,
        detail: impl Into<String>,
        digest: impl Into<String>,
    ) {
        if ok {
            self.pass_local(group, op, id, digest);
        } else {
            self.fail(group, op, id, detail);
        }
    }

    pub fn skip(&mut self, op: &str, workload: &str, reason: &str) {
        self.unsupported.push(Unsupported {
            op: op.to_string(),
            workload: workload.to_string(),
            reason: reason.to_string(),
        });
    }
}

fn cpu_time() -> (i64, i64) {
    // SAFETY: getrusage fills a plain POD struct we own.
    let mut ru: libc::rusage = unsafe { std::mem::zeroed() };
    if unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut ru) } != 0 {
        return (0, 0);
    }
    let ns = |t: libc::timeval| t.tv_sec * 1_000_000_000 + t.tv_usec * 1_000;
    (ns(ru.ru_utime), ns(ru.ru_stime))
}

fn max_rss_kib() -> i64 {
    let mut ru: libc::rusage = unsafe { std::mem::zeroed() };
    if unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut ru) } != 0 {
        return 0;
    }
    ru.ru_maxrss as i64
}

/// Runs one case: `warmup` unmeasured repetitions followed by `reps` measured
/// ones.
fn run_case(env: &Env, rec: &mut Recorder, c: &Case) {
    let p: &Profile = &env.profile;
    let mut shared = if c.per_rep {
        None
    } else {
        Some((c.setup)(env))
    };

    for _ in 0..p.warmup {
        match shared.as_mut() {
            Some(s) => rec.sink = rec.sink.wrapping_add((c.run)(env, s)),
            None => {
                let mut s = (c.setup)(env);
                rec.sink = rec.sink.wrapping_add((c.run)(env, &mut s));
                if let Some(f) = &c.free {
                    f(env, s);
                }
            }
        }
    }

    for rep in 0..p.reps {
        let mut owned = if c.per_rep {
            Some((c.setup)(env))
        } else {
            None
        };
        let state: &mut State = match owned.as_mut() {
            Some(s) => s,
            None => shared.as_mut().expect("shared state"),
        };
        let (u0, s0) = cpu_time();
        let t0 = Instant::now();
        let v = (c.run)(env, state);
        let wall = t0.elapsed();
        let (u1, s1) = cpu_time();
        rec.sink = rec.sink.wrapping_add(v);
        rec.samples.push(Sample {
            group: c.group.to_string(),
            op: c.op.clone(),
            workload: c.workload.clone(),
            threads: c.threads,
            rep,
            ops: c.ops,
            bytes: c.bytes,
            wall_ns: wall.as_nanos() as i64,
            cpu_user_ns: u1 - u0,
            cpu_sys_ns: s1 - s0,
            max_rss_kib: max_rss_kib(),
            status: "ok".to_string(),
        });
        if let Some(s) = owned
            && let Some(f) = &c.free
        {
            f(env, s);
        }
    }
    if let Some(s) = shared
        && let Some(f) = &c.free
    {
        f(env, s);
    }
}

/// Runs the registry in a stable order, so two runs of the same profile touch
/// the machine in the same sequence.
pub fn run_all(env: &Env, rec: &mut Recorder, mut cases: Vec<Case>) {
    cases.sort_by(|a, b| (a.group, &a.op, &a.workload).cmp(&(b.group, &b.op, &b.workload)));
    let trace = std::env::var_os("AMBER_CORE_OPS_TRACE").is_some();
    for c in &cases {
        let t0 = Instant::now();
        run_case(env, rec, c);
        if trace {
            eprintln!(
                "  {:<34} {:<28} {:8.1} ms",
                c.op,
                c.workload,
                t0.elapsed().as_secs_f64() * 1e3
            );
        }
    }
}

pub const SAMPLE_SCHEMA: &str = "amber-core-ops/samples/1";
